### ISSUE TITLE
Infinite loop in layout when there are no groups

### ISSUE THREAD

**Issue by domoritz** [vega/vega-view-transforms#1]
https://github.com/vega/vega-view-transforms/blob/master/src/gridLayout.js#L293 leads to an infinite loop when there are no groups. This results in the browser becoming unresponsive when I don't have any data. 

To trigger the bug, try any of the Vega-Lite example that use layout and use an invalid dataset url. 

For example, open this spec in the editor and it will freeze

```json
{
  "$schema": "https://vega.github.io/schema/vega-lite/v2.json",
  "data": {"url": "cars.json"},
  "mark": "point",
  "encoding": {
    "x": {
      "field": "Horsepower",
      "type": "quantitative"
    },
    "row": {"field": "Origin","type": "nominal"}
  }
}
```

**Comment by kanitw** [comment]
(Copied from slack) here is the Vega spec that produces the same issue:

```
{
  "$schema": "https://vega.github.io/schema/vega/v3.0.json",
  "autosize": "pad",
  "padding": 5,
  "data": [
    {
      "name": "source_0",
      "url": "cars.json",
      "format": {"type": "json","parse": {"Horsepower": "number"}},
      "transform": [
        {
          "type": "filter",
          "expr": "datum[\"Horsepower\"] !== null && !isNaN(datum[\"Horsepower\"])"
        }
      ]
    },
    {
      "name": "row_domain",
      "source": "source_0",
      "transform": [{"type": "aggregate","groupby": ["Origin"]}]
    }
  ],
  "signals": [
    {"name": "child_width","update": "200"},
    {"name": "child_height","update": "21"}
  ],
  "layout": {
    "padding": {"row": 10,"column": 10},
    "offset": 10,
    "columns": 1,
    "bounds": "full",
    "align": "all"
  },
  "marks": [
    {
      "name": "row_title",
      "role": "row-title",
      "type": "group",
      "marks": [
        {
          "type": "text",
          "role": "row-title-text",
          "style": "guide-title",
          "encode": {
            "update": {
              "align": {"value": "center"},
              "text": {"value": "Origin"},
              "angle": {"value": 270}
            }
          }
        }
      ]
    },
    {
      "name": "row_header",
      "type": "group",
      "role": "row-header",
      "from": {"data": "row_domain"},
      "sort": {"field": "datum[\"Origin\"]","order": "ascending"},
      "title": {
        "text": {"signal": "''+parent[\"Origin\"]"},
        "offset": 10,
        "orient": "left",
        "style": "guide-label",
        "encode": {
          "update": {
            "fontWeight": {"value": "normal"},
            "angle": {"value": 0},
            "fontSize": {"value": 10},
            "align": {"value": "right"},
            "baseline": {"value": "middle"}
          }
        }
      },
      "encode": {"update": {"height": {"signal": "child_height"}}}
    },
    {
      "name": "column_footer",
      "type": "group",
      "role": "column-footer",
      "encode": {"update": {"width": {"signal": "child_width"}}},
      "axes": [
        {
          "scale": "x",
          "labelOverlap": true,
          "orient": "bottom",
          "tickCount": {"signal": "ceil(child_width/40)"},
          "title": "Horsepower",
          "zindex": 1
        }
      ]
    },
    {
      "name": "cell",
      "type": "group",
      "style": "cell",
      "from": {
        "facet": {
          "name": "facet",
          "data": "source_0",
          "groupby": ["Origin"]
        }
      },
      "sort": {
        "field": ["datum[\"Origin\"]"],
        "order": ["ascending"]
      },
      "encode": {
        "update": {
          "width": {"signal": "child_width"},
          "height": {"signal": "child_height"}
        }
      },
      "marks": [
        {
          "name": "child_marks",
          "type": "symbol",
          "style": ["point"],
          "from": {"data": "facet"},
          "encode": {
            "update": {
              "x": {"scale": "x","field": "Horsepower"},
              "y": {"signal": "child_height","mult": 0.5},
              "stroke": {"value": "#4c78a8"},
              "fill": {"value": "transparent"},
              "opacity": {"value": 0.7}
            }
          }
        }
      ],
      "axes": [
        {
          "scale": "x",
          "domain": false,
          "grid": true,
          "labels": false,
          "maxExtent": 0,
          "minExtent": 0,
          "orient": "bottom",
          "tickCount": {"signal": "ceil(child_width/40)"},
          "ticks": false,
          "zindex": 0
        }
      ]
    }
  ],
  "scales": [
    {
      "name": "x",
      "type": "linear",
      "domain": {"data": "source_0","field": "Horsepower"},
      "range": [0,{"signal": "child_width"}],
      "nice": true,
      "zero": true
    }
  ],
  "config": {"axisY": {"minExtent": 30}}
}
```

**Comment by nyurik** [comment]
I would say this is the most frighting bug for me - if a browser freezes, and the vega editor auto-stores vega spec in the session store, reloading causes re-freeze, until the user clears their session store. Any ETA on it? No pressure of course ;)
P.S. this bug is making me rethink how nodejs service should detect and restart itself.

**Comment by jheer** [comment]
Fixed in 4a4e2164587f2f91162afa2570266f26d1811cd2, will be included in the next module and Vega releases.
